pkg/ipam/sync_test.go

```diff
@@ -0,0 +1,175 @@
+		now               time.Time
+		expiredThreshold  time.Time
+		expired           time.Time
+
+		fakeIpamServer *LiqoIPAM
+
+		addNetowrkToCache = func(ipamServer *LiqoIPAM, cidr string, creationTimestamp time.Time) {
+			ipamServer.cacheNetworks[cidr] = networkInfo{
+				cidr:              cidr,
+				creationTimestamp: creationTimestamp,
+			}
+		}
+
+		addIPToCache = func(ipamServer *LiqoIPAM, ip, cidr string, creationTimestamp time.Time) {
+			ipC := ipCidr{ip: ip, cidr: cidr}
+			ipamServer.cacheIPs[ipC.String()] = ipInfo{
+				ipCidr:            ipC,
+				creationTimestamp: creationTimestamp,
+			}
+		}
+
+		newNetwork = func(name, cidr string) *ipamv1alpha1.Network {
+			return &ipamv1alpha1.Network{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: testNamespace,
+				},
+				Spec: ipamv1alpha1.NetworkSpec{
+					CIDR: networkingv1beta1.CIDR(cidr),
+				},
+				Status: ipamv1alpha1.NetworkStatus{
+					CIDR: networkingv1beta1.CIDR(cidr),
+				},
+			}
+		}
+
+		newIP = func(name, ip, cidr string) *ipamv1alpha1.IP {
+			return &ipamv1alpha1.IP{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: testNamespace,
+				},
+				Spec: ipamv1alpha1.IPSpec{
+					IP:   networkingv1beta1.IP(ip),
+					CIDR: ptr.To(networkingv1beta1.CIDR(cidr)),
+				},
+				Status: ipamv1alpha1.IPStatus{
+					IP:   networkingv1beta1.IP(ip),
+					CIDR: networkingv1beta1.CIDR(cidr),
+				},
+			}
+		}
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		fakeClientBuilder = fake.NewClientBuilder().WithScheme(scheme.Scheme)
+		now = time.Now()
+		expiredThreshold = now.Add(-syncFrequency)
+		expired = now.Add(-2 * syncFrequency)
+	})
+
+	Describe("Testing the sync routine", func() {
+		Context("Sync Networks", func() {
+			BeforeEach(func() {
+				// Add in-cluster networks
+				client := fakeClientBuilder.WithObjects(
+					newNetwork("net1", "10.0.0.0/16"),
+					newNetwork("net2", "10.1.0.0/16"),
+					newNetwork("net3", "10.2.0.0/16"),
+				).Build()
+
+				// Populate the cache
+				fakeIpamServer = &LiqoIPAM{
+					Client:        client,
+					cacheNetworks: make(map[string]networkInfo),
+				}
+				addNetowrkToCache(fakeIpamServer, "10.0.0.0/16", now)
+				addNetowrkToCache(fakeIpamServer, "10.1.0.0/16", expired)
+				addNetowrkToCache(fakeIpamServer, "10.3.0.0/16", expired)
+				addNetowrkToCache(fakeIpamServer, "10.4.0.0/16", now)
+			})
+
+			It("should remove networks from cache if they are not present in the cluster", func() {
+				// Run sync
+				Expect(fakeIpamServer.syncNetworks(ctx, expiredThreshold)).To(Succeed())
+
+				// Check the cache
+				Expect(fakeIpamServer.cacheNetworks).To(HaveKey("10.0.0.0/16"))    // network in cluster and cache
+				Expect(fakeIpamServer.cacheNetworks).To(HaveKey("10.1.0.0/16"))    // network in cluster and cache before expired threshold
+				Expect(fakeIpamServer.cacheNetworks).NotTo(HaveKey("10.2.0.0/16")) // network in cluster but not in cache
```

Comment:
Shouldn't it add the network when available in the cluster since the cluster is the source of truth?